notes/trees-that-grow-proc-macro.md

```diff
@@ -0,0 +1,167 @@
+As nickel grows, it becomes unwieldy to use a single AST for every stage of compilation. Some things are kept out of the AST, even though they would be useful in some places, because they don't make sense and would get in the way at others (e.g. types for the LSP, closures at runtime). Other things are in the AST, but aren't actually filled in until later on in the process (e.g. forall's var_kind). This causes `unwrap()`s and `unreachable!()`s and hurts code maintainability.
+
+## Strategies
+
+One strategy to deal with this is to repeat a large AST enum in multiple places with small differences, and convert from one to another as we move through compilation. This creates a large maintenance burden, since many changes have to be propagated to all the trees.
+
+Another strategy is outlined by the paper [Trees that Grow](https://www.microsoft.com/en-us/research/uploads/prod/2016/11/trees-that-grow.pdf). This describes an extendable tree structure using type families and pattern synonyms for ergonomics. While this is quite a bit DRYer, there is still a decent amount of boilerplate, and actually using these trees is less intuitive. Especially in rust, where there are no pattern synonyms.
+
+I (@Radvendii) have explored using proc macros to get the best of both worlds, and remove the boilerplate along the way. There are three ways we could use proc macros, but they share a common syntax.
+
+## Macro Syntax
+
+The naming of things is up in the air, but the general idea is to annotate the tree definition with how fields get modified in the extensions. There are two kinds of extensions, product extensions which add data to an existing term type, and sum extensions which add a new term type.
+
+```rust
```

Comment:
It could help to add a written description of what the following program is supposed to mean, that is:

> The following example models 3 different ASTs, such that ...

It's not obvious what `Foo` is here (a type? a name? something else), and it's a bit hard to follow the translation if we don't really know what we are translating to begin with.